### PR TITLE
Add support for Hyundai Tucson 2021

### PR DESCRIPTION
**Pull request by bluesforte**
This is a working port for the Hyundai Tucson 2021. It has been tested by one of my clients on multiple long drives (3+hours each), and seems to be working well. I didn't do any lateral PID tune due to time constraints, but the default one handles pretty well.

Some notes:

- SteerRatio was determined empirically - 16 "felt" right after a few test drives.
- Car data (wheelbase, weight etc) was determined from the Hyundai website
- The Tucson cruise control appears not to turn on unless above ~20MPH, though does correctly stay on and allows for following cars even in stop-and-go traffic. May be able to be overridden somewhere in openpilot, but minimizing changes and don't want to have any impact on the safety model.
- Wish I had saved more test data while I had access to the car, but wasn't thinking ahead to the pull request :(. If I do get access again, I will grab the route data to add to test_routes.py.

Enjoy!

### CODE REVIEW

File: selfdrive/car/hyundai/values.py [hunk 1]
```diff
@@ -41,6 +41,7 @@ class CAR:
   SANTA_FE_2022 = "HYUNDAI SANTA FE 2022"
   SONATA = "HYUNDAI SONATA 2020"
   SONATA_LF = "HYUNDAI SONATA 2019"
+  TUCSON_2021 = "HYUNDAI TUCSON_2021"
```
Comment (adeebshihadeh) [inline selfdrive/car/hyundai/values.py:44]:
```suggestion
  TUCSON = "HYUNDAI TUCSON 2021"
```

File: selfdrive/car/hyundai/values.py [hunk 2]
```diff
@@ -156,6 +157,9 @@ class Buttons:
...
   }],
+  CAR.TUCSON_2021: [{
+    64: 8, 66: 8, 67: 8, 68: 8, 127: 8, 273: 8, 274: 8, 275: 8, 339: 8, 356: 4, 399: 8, 447: 8, 512: 6, 544: 8, 593: 8, 608: 8, 688: 5, 790: 8, 809: 8, 832: 8, 884: 8, 897: 8, 899: 8, 902: 8, 903: 8, 909: 8, 916: 8, 1040: 8, 1056: 8, 1057: 8, 1064: 8, 1065: 8, 1078: 4, 1151: 6, 1168: 7, 1170: 8, 1186: 2, 1191: 2, 1253: 8, 1254: 8, 1255: 8, 1265: 4, 1280: 1, 1282: 4, 1287: 4, 1290: 8, 1292: 8, 1294: 8, 1312: 8, 1313: 8, 1314: 8, 1322: 8, 1342: 6, 1345: 8, 1348: 8, 1349: 8, 1351: 8, 1353: 8, 1363: 8, 1365: 8, 1366: 8, 1367: 8, 1369: 8, 1371: 8, 1407: 8, 1415: 8, 1419: 8, 1427: 6, 1440: 8, 1456: 4, 1470: 8, 1472: 8, 1486: 8, 1487: 8, 1491: 8, 1530: 8, 1532: 5, 1792: 8, 1810: 8, 1813: 8, 1872: 8, 1879: 8, 1937: 8, 1953: 8, 1968: 8, 1988: 8, 1996: 8, 2000: 8, 2001: 8, 2004: 8, 2008: 8, 2009: 8, 2012: 8, 2015: 8, 2016: 8, 2017: 8, 2024: 8, 2025: 8
+  }]
```
Comment (adeebshihadeh) [inline selfdrive/car/hyundai/values.py:162]: CAN fingerprinting is deprecated
```suggestion
```

File: selfdrive/car/hyundai/values.py [hunk 3]
```diff
@@ -532,6 +536,20 @@ class Buttons:
       b"\xf1\x87LBLUFN622950KF36\xa8\x88\x88\x88\x87w\x87xh\x99\x96\x89\x88\x99\x98\x89\x88\x99\x98\x89\x87o\xf6\xff\x98\x88o\xffx'\xf1\x81U891\x00\x00\x00\x00\x00\x00\xf1\x00bcsh8p54  U891\x00\x00\x00\x00\x00\x00SLX2G38NB3\xd1\xc3\xf8\xa8",
     ],
   },
+  CAR.TUCSON_2021: {
```
Comment (adeebshihadeh) [inline selfdrive/car/hyundai/values.py:539]:
```suggestion
  CAR.TUCSON: {
```

File: selfdrive/car/hyundai/values.py [hunk 4]
```diff
@@ -861,7 +879,7 @@ class Buttons:
 FEATURES = {
   # which message has the gear
   "use_cluster_gears": set([CAR.ELANTRA, CAR.ELANTRA_GT_I30, CAR.KONA]),
-  "use_tcu_gears": set([CAR.KIA_OPTIMA, CAR.SONATA_LF, CAR.VELOSTER]),
+  "use_tcu_gears": set([CAR.KIA_OPTIMA, CAR.SONATA_LF, CAR.VELOSTER, CAR.TUCSON_2021]),
```
Comment (adeebshihadeh) [inline selfdrive/car/hyundai/values.py:882]:
```suggestion
  "use_tcu_gears": set([CAR.KIA_OPTIMA, CAR.SONATA_LF, CAR.VELOSTER, CAR.TUCSON]),
```

File: selfdrive/car/hyundai/values.py [hunk 5]
```diff
@@ -904,6 +922,7 @@ class Buttons:
   CAR.SANTA_FE_2022: dbc_dict('hyundai_kia_generic', None),
   CAR.SONATA: dbc_dict('hyundai_kia_generic', 'hyundai_kia_mando_front_radar'),
   CAR.SONATA_LF: dbc_dict('hyundai_kia_generic', None), # Has 0x5XX messages, but different format
+  CAR.TUCSON_2021: dbc_dict('hyundai_kia_generic', 'hyundai_kia_mando_front_radar'),
```
Comment (adeebshihadeh) [inline selfdrive/car/hyundai/values.py:925]:
```suggestion
  CAR.TUCSON: dbc_dict('hyundai_kia_generic', None),
```

File: selfdrive/car/hyundai/interface.py [hunk 6]
```diff
@@ -143,6 +143,15 @@ def get_params(candidate, fingerprint=gen_empty_fingerprint(), car_fw=[]):  # py
       tire_stiffness_factor = 0.5
       ret.lateralTuning.pid.kiBP, ret.lateralTuning.pid.kpBP = [[0.], [0.]]
       ret.lateralTuning.pid.kpV, ret.lateralTuning.pid.kiV = [[0.25], [0.05]]
+    elif candidate == CAR.TUCSON_2021:
```
Comment (adeebshihadeh) [inline selfdrive/car/hyundai/interface.py:146]:
```suggestion
    elif candidate == CAR.TUCSON:
```

**Review by adeebshihadeh — CHANGES_REQUESTED** [review]
Thanks for the car port, but we can't merge this without a test route. Are you sure you don't have any uploaded routes from the Tucson? If you still have the comma two or three, then the routes are still likely on the device.

**Comment by bluesforte** [comment]
> 
> 
> Thanks for the car port, but we can't merge this without a test route. Are you sure you don't have any uploaded routes from the Tucson? If you still have the comma two or three, then the routes are still likely on the device.

Certainly understand.  I don't have access to the device anymore as it was an install for a friend's car and I'm away at the moment.  Next time I meet up with him I'll grab the latest routes off the device.  Just wanted to get this submitted here in case it helps anyone else.

**Comment by sshane** [comment]
@bluesforte any updates on getting a test route?

**Comment by sshane** [comment]
Moved to https://github.com/commaai/openpilot/pull/24791 to resolve conflicts, couldn't push. Thanks for the PR!

**Comment by bluesforte** [comment]
Ah sorry I never replied - yeah I wasn't able to get the car or device again to pull the logs :(.  Sorry.  Hopefully the PR is good enough for anyone else with a Tucson!

**Comment by sshane** [comment]
@bluesforte Two quick questions, do you remember if the harness was Hyundai L? And when you say it doesn't engage below 20, you mean you can't engage it when you're physically under 20 mph, or that the set speed is limited to 20 mph? The latter is normal

**Comment by bluesforte** [comment]
I remember when we had bought the harness we had gone with the one the site 'thought' might work (can't seem to find the page I remember from back then).  I do believe it was an L though.

Re: engaging, I meant the former actually.  In order to engage it the first time, the vehicle needed to be moving >~20mph.  After that, it'd do stop-and-go just fine.  It just wouldn't let you engage it from a stop or when moving too slowly.

**Comment by sshane** [comment]
Yeah SCC normally won't let you engage under 5 mph without a lead, and 20 is the minimum set speed of cruise control (you engage at 6 mph, it speeds up to 20 for example).

There might be some qlogs I can check for this. Thanks for confirming the harness!

**Comment by bluesforte** [comment]
> You had to be above 20 mph every time to engage, not just the first time?

Yep.  Min speed was indeed 20mph as well.  The part I specifically remember being strange was that the speed at which it engaged seemed to vary slightly, and I wasn't sure what it was based on.  Sometimes it'd engage at ~15mph, other times at ~20mph.  Didn't seem to be a fixed limit (though maybe it was a time-delayed gate or something).

**Pull request closed without merging**